### ISSUE TITLE
Prefix operator expression printed incorrectly

### ISSUE THREAD

**Issue by dtonhofer** [SWI-Prolog/swipl-devel#761]
In version 8.3.16, trying out operators:

We define:

```
?-
op(500,fy,⊓).
```

Let's see the interwork with `+`:

`+` has precedence value, 500, and is `yfx`

The following case is correctly parsed and printed:

```
?- 
X = ⊓ ⊓ 4 + 5, write_canonical(X).

+(⊓(⊓(4)),5)
X = ⊓ ⊓4+5.
```

Parenthesizing the `+` subexpression, the following case is correctly parsed but
is printed **the same as above**. That's ... not correct.

```
?- 
X = ⊓ ⊓ (4 + 5), write_canonical(X).

⊓(⊓(+(4,5)))
X = ⊓ ⊓4+5.    % This looks like a bug, should be ⊓ ⊓(4+5)
```

Re-read the output of above. The result is as expected `+(⊓(⊓(4)),5)`, not `⊓(⊓(+(4,5)))`.

```
?- 
X = ⊓ ⊓4+5, write_canonical(X).
+(⊓(⊓(4)),5)
X = ⊓ ⊓4+5.
```

Try with higher precedence (lower precedence value) on the right:

Parsing is correct with `*`, which has has precedence value, 400 (i.e. binds more strongly than `⊓`), type `yfx` 

```
?- 
X = ⊓ ⊓ 4 * 5, write_canonical(X).

⊓(⊓(*(4,5)))
X = ⊓ ⊓4*5.
```

```
?- X = ⊓ ⊓ (4 * 5), write_canonical(X).

⊓(⊓(*(4,5)))
X = ⊓ ⊓4*5.
```

Re-parse the output above:

```
?- X = ⊓ ⊓4*5, write_canonical(X).

⊓(⊓(*(4,5)))
X = ⊓ ⊓4*5.
```

In this case, perfect.

**Comment by JanWielemaker** [comment]
It is indeed not good.   On the other hand this is basically ambiguous as both ⊓ and + winning satisfies the operator constraints.   I think this is not allowed in ISO.   SWI-Prolog's parser is more relaxed than ISO.  SICStus Prolog parser is even more relaxed, doing a search process that commits to the first solution.   That too results in various non-intuitive results.  I've had some discussion with SICStus on this matter.   Despite their strict adherence to the ISO standard they had concluded that their users did not like enforcing the restricted ISO interpretation of operators.

Bottom line: make sure the operators you intend to combine are not ambiguous.  Then you will still be faced with a few cases where the SWI-Prolog parser may reject thing that you'd expect to be accepted.  SICStus will accept them all and ISO will reject more.

My recent encounters with this stuff have learned me that is is probably unwise to modify anything or (eventually) adhere the SICStus model.

**Comment by dtonhofer** [comment]
Thanks Jan. This must be where art meets parsing.

But when printing these out, do I have to parenthesize manually to eliminate ambiguity? write_canonical/1 generates an unambiguous representation, but is there some built-in to have `⊓(⊓(+(4,5)))` printed as `⊓ ⊓(4+5).` instead of `⊓ ⊓4+5.` or do I have to code something (more art! then one could also check whether the printed representation parse back into the original tree before putting it on screen ... )

**Comment by JanWielemaker** [comment]
I guess the best solution is to define ⊓ just above or just below +.   (Why) is that a problem?

**Comment by dtonhofer** [comment]
I lost sight of the fact that you can set the precedence value too ...

Well, it's not a problem really. I was just poking around at the edges and writing a README on `op/3`. 

Let's close this!
